### PR TITLE
CI/CD Improvements

### PR DESCRIPTION
**Pull request by JacobMGEvans**
# CI/CD Improvements
## Changeset
Adding configuration allows for use of CLI for changesets. A necessary supplement to the changesets bot, and GitHub Action. 
- Installed Changeset CLI tool
- NPX changeset init
    - Added changesets directory
    - Config
    - README
- Modified the config for `main` branch instead of `master`

## ESLint & Prettier Integration
Running Prettier as a rule through ESLint to improve CI/CD usage
- Added additional TypeScript support for ESLint 
- Prettier errors as ESLint rule
- .vscode directory w/ settings.json config added that enforces 
the usage of ESLint by anyone working in the workspace

### CODE REVIEW

**Comment by JacobMGEvans** [comment]
@threepointone  I will rebase all the commits later and the HEAD

**Comment by JacobMGEvans** [comment]
GitHub Action Workflows:

- [ x ] Test, 
- [ x ] Linting 
- [ x ] Type Checking 

The Test & Typechecking workflow `success` is required for the Release Workflow to be capable of running.

File: .github/notification.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const path = require("path");
+const require = createRequire(import.meta.url);
```
Comment (threepointone) [inline .github/notification.js:2]: what is createRequire? this repo isn't defined as a `type: "module"`, so would it be required here?

File: .github/notification.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+async function run() {
+  const releases = process.argv.slice(2)[0];
+  const data = JSON.parse(releases);
+  const packages = await Promise.all(
```
Comment (threepointone) [inline .github/notification.js:7]: there's nothing async happening here, so why is this using Promise.all?

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "editor.codeActionsOnSave": {
+    "source.fixAll.eslint": true
```
Comment (threepointone) [inline .vscode/settings.json:3]: let's not do a fix on save. in my experience, it gets confusing.

File: .github/workflows/tests-types.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
+
+# Prefer offline will skip registry verification with cached NPM packages
+      - name: Install NPM dependencies
+        run: npm ci --prefer-offline
+      
+      - name: Run Type Checking & ESLint
+        run: tsc --skipLibCheck && npm run lint
```
Comment (threepointone) [inline .github/workflows/tests-types.yml:58]: add a TODO comment here to remove the `--skipLibCheck` flag at some point

File: .github/workflows/releases.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NPM_PUBLISH_TOKEN: ${{ secrets.NPM_PUBLISH_TOKEN }}
+
+      - name: Create Notification
+        if: steps.changesets.outputs.published == 'true'
+        id: notification
+        run: message=$(node notification.js '${{ steps.changesets.outputs.publishedPackages }}') && echo ::set-output name=message::${message//$'\n'/'%0A'}
+
+
+      - name: Discord Message
```
Comment (threepointone) [inline .github/workflows/releases.yml:48]: let's do the discord stuff in a separate PR

File: .github/workflows/prereleases.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# name: Prerelease
```
Comment (threepointone) [inline .github/workflows/prereleases.yml:1]: why don't we keep this around? looks like it'll still work, sans changelogs, which is fine.

File: .github/workflows/releases.yml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+      - name: Install Dependencies
+        run: npm ci --prefer-offline
+
+      - name: Create Release Pull Request or Publish to npm
+        id: changesets
+        uses: changesets/action@v1
+        with:
+          # This expects you to have a script called release which does a build for your packages and calls changeset publish
+          publish: npm changeset:publish
```
Comment (threepointone) [inline .github/workflows/releases.yml:37]: I would much rather this be an explicit call to changesets, with a `--tag beta`, instead of calling an npm script.

File: .github/workflows/tests-types.yml [hunk 8]
```diff
@@ -0,0 +1,62 @@
+name: CI
```
Comment (threepointone) [inline .github/workflows/tests-types.yml:1]: maybe we can do the tests/types stuff in a separate pr too, and we can land this one quickly.

**Review by threepointone — COMMENTED** [review]
started leaving some notes

File: .vscode/settings.json (hunk 3 above)
Comment (JacobMGEvans) [reply .vscode/settings.json:3]: I will switch it to a runAll or something similar

File: .github/workflows/releases.yml (hunk 5 above)
Comment (JacobMGEvans) [reply .github/workflows/releases.yml:48]: Ah! Cool that gives me more time to figure out the notification file logic

File: .github/workflows/prereleases.yml (hunk 6 above)
Comment (JacobMGEvans) [reply .github/workflows/prereleases.yml:1]: I can not really have only release or prerelease run based on the current Event trigger which is `push to main`

File: .github/workflows/releases.yml (hunk 7 above)
Comment (JacobMGEvans) [reply .github/workflows/releases.yml:37]: Sounds good

File: .github/workflows/releases.yml [hunk 9]
```diff
@@ -0,0 +1,39 @@
+name: Release
```
Comment (threepointone) [inline .github/workflows/releases.yml:1]: remove

File: .eslintrc [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    "./tsconfig.json",
+    "eslint:recommended",
+    "plugin:@typescript-eslint/recommended"
+  ],
+  "plugins": ["@typescript-eslint", "prettier"],
+  "rules": {
+    "prettier/prettier": "error"
+  },
+  "includes": [
+    "./**/*.ts",
```
Comment (threepointone) [inline .eslintrc:17]: also include tsx and jsx files

File: .eslintrc [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    "./**/*.js"
+  ],
+  "excludes": [
+    "./**/*.d.ts",
+    "./node_modules"
+  ]
+}
```
Comment (threepointone) [inline .eslintrc:24]: add a newline

File: .vscode/extensions.json [hunk 12]
```diff
@@ -0,0 +1,5 @@
+{
+  "recommendations": [
+    "dbaeumer.vscode-eslint"
```
Comment (threepointone) [inline .vscode/extensions.json:3]: and prettier

File: .vscode/settings.json [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
+  "eslint.format.enable": true,
+  "[javascript]": { "editor.defaultFormatter": "dbaeumer.vscode-eslint" },
+  "[typescript]": { "editor.defaultFormatter": "dbaeumer.vscode-eslint" }
+}
```
Comment (threepointone) [inline .vscode/settings.json:5]: the default formatter should be prettier

Comment (JacobMGEvans) [reply .vscode/settings.json:5]: We are running Prettier as a Rule through ESLint

Comment (threepointone) [reply .vscode/settings.json:5]: yeah, but won't running eslint as a formatter also "fix" issues automatically? I'd like to avoid that implicit behaviour

**Review by threepointone — APPROVED** [review]
Stamping this, take it for a spin

**Review by Spiritaine — COMMENTED** [review]
Nice

**Review by Spiritaine — COMMENTED** [review]
Nice
